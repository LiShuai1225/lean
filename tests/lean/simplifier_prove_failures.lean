open tactic

constants (P Q R : Prop) (HP : P) (HPQ : P → Q) (HQR : Q → R = true)
attribute HQR [simp]

meta_definition prove_skip           : tactic unit := skip
meta_definition prove_fail           : tactic unit := failed
meta_definition prove_partial_assign : tactic unit := mk_const `HPQ >>= apply
meta_definition prove_full_assign    : tactic unit := (mk_const `HPQ >>= apply) >> (mk_const `HP) >>= exact

set_option trace.simplifier.prove true

example : R := by simplify_goal prove_skip []           >> try triv
example : R := by simplify_goal prove_fail []           >> try triv
example : R := by simplify_goal prove_partial_assign [] >> try triv
example : R := by simplify_goal prove_full_assign []    >> try triv
