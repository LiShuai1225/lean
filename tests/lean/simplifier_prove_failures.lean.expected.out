[simplifier.prove] goal: ?m_1 : Q
[simplifier.prove] prove_fn succeeded but did not return a proof
simplifier_prove_failures.lean:13:15: error: simp tactic failed to simplify
state:
⊢ R
simplifier_prove_failures.lean:13:15: error: failed to add declaration 'example' to environment, value has metavariables
remark: set 'formatter.hide_full_terms' to false to see the complete term
  ?M_1
[simplifier.prove] goal: ?m_1 : Q
[simplifier.prove] prove_fn failed to prove Q
simplifier_prove_failures.lean:14:15: error: simp tactic failed to simplify
state:
⊢ R
simplifier_prove_failures.lean:14:15: error: failed to add declaration 'example' to environment, value has metavariables
remark: set 'formatter.hide_full_terms' to false to see the complete term
  ?M_1
[simplifier.prove] goal: ?m_1 : Q
[simplifier.prove] prove_fn succeeded but left an unrecognized metavariable of type P in proof
simplifier_prove_failures.lean:15:15: error: simp tactic failed to simplify
state:
⊢ R
simplifier_prove_failures.lean:15:15: error: failed to add declaration 'example' to environment, value has metavariables
remark: set 'formatter.hide_full_terms' to false to see the complete term
  ?M_1
[simplifier.prove] goal: ?m_1 : Q
[simplifier.prove] success: HPQ HP : Q
